In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import re
import string
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords') # downloading stopwords from nltk library

    # Starting With GossipCop Datasets
df_gossip_fake = pd.read_csv("gossipcop_fake.csv")
df_gossip_real = pd.read_csv("gossipcop_real.csv")

df_gossip_fake.info()




[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5323 entries, 0 to 5322
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         5323 non-null   object
 1   news_url   5067 non-null   object
 2   title      5323 non-null   object
 3   tweet_ids  5135 non-null   object
dtypes: object(4)
memory usage: 166.5+ KB


In [ ]:
df_gossip_fake.head(10)

,id,news_url,title,tweet_ids
0,gossipcop-2493749932,www.dailymail.co.uk/tvshowbiz/article-5874213/...,Did Miley Cyrus and Liam Hemsworth secretly ge...,284329075902926848\t284332744559968256\t284335...
1,gossipcop-4580247171,hollywoodlife.com/2018/05/05/paris-jackson-car...,Paris Jackson & Cara Delevingne Enjoy Night Ou...,992895508267130880\t992897935418503169\t992899...
2,gossipcop-941805037,variety.com/2017/biz/news/tax-march-donald-tru...,Celebrities Join Tax March in Protest of Donal...,853359353532829696\t853359576543920128\t853359...
3,gossipcop-2547891536,www.dailymail.co.uk/femail/article-3499192/Do-...,Cindy Crawford's daughter Kaia Gerber wears a ...,988821905196158981\t988824206556172288\t988825...
4,gossipcop-5476631226,variety.com/2018/film/news/list-2018-oscar-nom...,Full List of 2018 Oscar Nominations – Variety,955792793632432131\t955795063925301249\t955798...
5,gossipcop-5189580095,www.townandcountrymag.com/society/tradition/a1...,Here's What Really Happened When JFK Jr. Met P...,890253005299351552\t890401381814870016\t890491...
6,gossipcop-9588339534,www.foxnews.com/entertainment/2016/12/16/bigge...,Biggest celebrity scandals of 2016,683226380742557696\t748604615503929345\t748604...
7,gossipcop-8753274298,www.eonline.com/news/958257/caitlyn-jenner-add...,Caitlyn Jenner Addresses Rumored Romance With ...,1026891446081728512\t1026891745219543043\t1026...
8,gossipcop-8105333868,www.inquisitr.com/3871816/taylor-swift-reporte...,Taylor Swift Reportedly Reacts To Tom Hiddlest...,818928533569437697\t819100640878202880\t819174...
9,gossipcop-2803748870,www.huffingtonpost.com/entry/kate-mckinnon-the...,"For The Love Of God, Why Can't Anyone Write Ka...",816030248190046212\t816030859484626947\t816049...


We have id , news_url (url of the news) , title of the news , tweet ids columns


In [ ]:
df_gossip_real.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16817 entries, 0 to 16816
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         16817 non-null  object
 1   news_url   16804 non-null  object
 2   title      16817 non-null  object
 3   tweet_ids  15759 non-null  object
dtypes: object(4)
memory usage: 525.7+ KB


In [ ]:
df_gossip_real.head(10)

,id,news_url,title,tweet_ids
0,gossipcop-882573,https://www.brides.com/story/teen-mom-jenelle-...,Teen Mom Star Jenelle Evans' Wedding Dress Is ...,912371411146149888\t912371528343408641\t912372...
1,gossipcop-875924,https://www.dailymail.co.uk/tvshowbiz/article-...,Kylie Jenner refusing to discuss Tyga on Life ...,901989917546426369\t901989992074969089\t901990...
2,gossipcop-894416,https://en.wikipedia.org/wiki/Quinn_Perkins,Quinn Perkins,931263637246881792\t931265332022579201\t931265...
3,gossipcop-857248,https://www.refinery29.com/en-us/2018/03/19192...,I Tried Kim Kardashian's Butt Workout & Am For...,868114761723936769\t868122567910936576\t868128...
4,gossipcop-884684,https://www.cnn.com/2017/10/04/entertainment/c...,Celine Dion donates concert proceeds to Vegas ...,915528047004209152\t915529285171122176\t915530...
5,gossipcop-899309,https://tech2.org/chris-evans-millie-bobby-bro...,"Chris Evans, Millie Bobby Brown, Snoop Dogg an...",940013737204899841\t940014043619831808\t940014...
6,gossipcop-848092,https://tvline.com/2018/05/02/the-handmaids-ta...,Handmaid's Tale Renewed for Season 3,859773820910977024\t859773844348837889\t859773...
7,gossipcop-897293,https://fashionmagazine.com/culture/selena-gom...,A Complete Timeline of Selena Gomez and Justin...,936299044422520832\t936301588624396290\t936301...
8,gossipcop-926957,https://decider.com/2018/06/11/claws-season-2-...,When Will ‘Claws’ Season 2 Be On Hulu?,984486548996419584\t984490492787351556\t984491...
9,gossipcop-930832,http://www.criticschoice.com/critics-choice-aw...,Critics' Choice Awards - Critics' Choice Awards,989840869443457024\t990987021194743808


In [ ]:
df_gossip_fake['label'] = 1  # labeling the fake data as 1
df_gossip_real['label'] = 0  # labeling the fake data as 0


In [ ]:
# now we use the same process and combine all csv files in one dataset
df_politi_fake = pd.read_csv("politifact_fake.csv")
df_politi_real = pd.read_csv("politifact_real.csv")

In [ ]:
df_politi_fake.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 432 entries, 0 to 431
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         432 non-null    object
 1   news_url   428 non-null    object
 2   title      432 non-null    object
 3   tweet_ids  392 non-null    object
dtypes: object(4)
memory usage: 13.6+ KB


In [ ]:
df_politi_real.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 624 entries, 0 to 623
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         624 non-null    object
 1   news_url   567 non-null    object
 2   title      624 non-null    object
 3   tweet_ids  409 non-null    object
dtypes: object(4)
memory usage: 19.6+ KB


In [ ]:
df_politi_fake['label'] = 1  # labeling politifacts
df_politi_real['label'] = 0

In [ ]:
nulls1 = df_gossip_fake.isnull().sum()   # to see null values
print(f"\n Total null values in gossipcop_fake.csv: {nulls1}")

nulls2 = df_gossip_real.isnull().sum()
print(f"Total null values in gossipcop_real.csv: {nulls2}")

nulls3 = df_politi_fake.isnull().sum()
print(f"Total null values in politifact_fake.csv: {nulls3}")

nulls4 = df_politi_real.isnull().sum()
print(f"Total null values in politifact_real.csv: {nulls4}")




 Total null values in gossipcop_fake.csv: id             0
news_url     256
title          0
tweet_ids    188
label          0
dtype: int64
Total null values in gossipcop_real.csv: id              0
news_url       13
title           0
tweet_ids    1058
label           0
dtype: int64
Total null values in politifact_fake.csv: id            0
news_url      4
title         0
tweet_ids    40
label         0
dtype: int64
Total null values in politifact_real.csv: id             0
news_url      57
title          0
tweet_ids    215
label          0
dtype: int64


In [ ]:
nulls1 = nulls1.sum()
nulls2 = nulls2.sum()
nulls3 = nulls3.sum()
nulls4 = nulls4.sum()

nulls5 = nulls1 + nulls2 + nulls3 + nulls4
print(f"Total number of null values in all datasets: {nulls5}")

Total number of null values in all datasets: 1831


In [ ]:
# combining all of them
# we overlook null values in news_url and tweet_ids to avoid data loss
df = pd.concat([df_gossip_fake, df_gossip_real, df_politi_fake, df_politi_real], ignore_index=True)

In [ ]:
df.head(10)

,id,news_url,title,tweet_ids,label
0,gossipcop-2493749932,www.dailymail.co.uk/tvshowbiz/article-5874213/Did-Miley-Cyrus-Liam-Hemsworth-secretly-married.html,Did Miley Cyrus and Liam Hemsworth secretly get married?,284329075902926848\t284332744559968256\t284335412590297089\t284359918792880128\t2843853851513323...,1
1,gossipcop-4580247171,hollywoodlife.com/2018/05/05/paris-jackson-cara-delevingne-matching-outfits-night-out-nyc-dating...,Paris Jackson & Cara Delevingne Enjoy Night Out In Matching Outfits: They Have ‘Amazing Chemistry’,992895508267130880\t992897935418503169\t992899529329569792\t992904299674923009\t9929063305905971...,1
2,gossipcop-941805037,variety.com/2017/biz/news/tax-march-donald-trump-protest-1202031487/,Celebrities Join Tax March in Protest of Donald Trump,853359353532829696\t853359576543920128\t853359758400729088\t853360019814973440\t8533600735648686...,1
3,gossipcop-2547891536,www.dailymail.co.uk/femail/article-3499192/Do-blondes-REALLY-fun-Cindy-Crawford-s-model-daughter...,Cindy Crawford's daughter Kaia Gerber wears a wig after dining with Harry Styles,988821905196158981\t988824206556172288\t988825130838077440\t988827402749919232\t9888354116098457...,1
4,gossipcop-5476631226,variety.com/2018/film/news/list-2018-oscar-nominations-1202668757/,Full List of 2018 Oscar Nominations – Variety,955792793632432131\t955795063925301249\t955798007861170178\t955801005643452418\t9558014527814082...,1
5,gossipcop-5189580095,www.townandcountrymag.com/society/tradition/a10241220/jfk-jr-princess-diana/,Here's What Really Happened When JFK Jr. Met Princess Diana,890253005299351552\t890401381814870016\t890491475363938305\t890504639896072193,1
6,gossipcop-9588339534,www.foxnews.com/entertainment/2016/12/16/biggest-celebrity-scandals-2016.html,Biggest celebrity scandals of 2016,683226380742557696\t748604615503929345\t748604615340355585\t748604615332007936\t7486051302918963...,1
7,gossipcop-8753274298,www.eonline.com/news/958257/caitlyn-jenner-addresses-rumored-romance-with-sophia-hutchins,Caitlyn Jenner Addresses Rumored Romance With Sophia Hutchins,1026891446081728512\t1026891745219543043\t1026892913308049408\t1026892997009584130\t102689334633...,1
8,gossipcop-8105333868,www.inquisitr.com/3871816/taylor-swift-reportedly-reacts-to-tom-hiddlestons-golden-globes-win/,Taylor Swift Reportedly Reacts To Tom Hiddleston’s Golden Globes Win,818928533569437697\t819100640878202880\t819174790930034690\t834689684538273793,1
9,gossipcop-2803748870,www.huffingtonpost.com/entry/kate-mckinnon-the-spy-who-dumped-me_us_5b61e93ae4b0fd5c73d59a48,"For The Love Of God, Why Can't Anyone Write Kate McKinnon A Good Movie Role?",816030248190046212\t816030859484626947\t816049169471250432\t816075827959959552\t8161395599017287...,1


In [ ]:
print("Data Loaded.")
print(f"Total Rows: {len(df)}")
df.info()

Data Loaded.
Total Rows: 23196
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23196 entries, 0 to 23195
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         23196 non-null  object
 1   news_url   22866 non-null  object
 2   title      23196 non-null  object
 3   tweet_ids  21695 non-null  object
 4   label      23196 non-null  int64 
dtypes: int64(1), object(4)
memory usage: 906.2+ KB


Here, fake and real dataframes are combined,at first fake news dataframe with 16817 rows,then real dataframe's 5323 rows until the end.

In [ ]:
stop_words = set(stopwords.words('english'))

In [ ]:
def preprocess_text(text):  # Preprocessing setup
  text = str(text).lower() # lowercase
  text = re.sub(f"[{re.escape(string.punctuation)}]", "", text) # remove punctuation
  tokens = text.split() # tokenize

  # Remove stopwords and short tokens
  cleaned_tokens = [t for t in tokens if t not in stop_words and len(t) > 2]
  return " ".join(cleaned_tokens)


print("Cleaning started , please wait...")
df["clean_title"] = df["title"].apply(preprocess_text) # creating new column with cleaned titles using our function

print("Before Cleaning / After Cleaning")
pd.set_option('display.max_colwidth', 100)
display(df[['title', 'clean_title', 'label']].head(10))

Cleaning started , please wait...
Before Cleaning / After Cleaning


,title,clean_title,label
0,Did Miley Cyrus and Liam Hemsworth secretly get married?,miley cyrus liam hemsworth secretly get married,1
1,Paris Jackson & Cara Delevingne Enjoy Night Out In Matching Outfits: They Have ‘Amazing Chemistry’,paris jackson cara delevingne enjoy night matching outfits ‘amazing chemistry’,1
2,Celebrities Join Tax March in Protest of Donald Trump,celebrities join tax march protest donald trump,1
3,Cindy Crawford's daughter Kaia Gerber wears a wig after dining with Harry Styles,cindy crawfords daughter kaia gerber wears wig dining harry styles,1
4,Full List of 2018 Oscar Nominations – Variety,full list 2018 oscar nominations variety,1
5,Here's What Really Happened When JFK Jr. Met Princess Diana,heres really happened jfk met princess diana,1
6,Biggest celebrity scandals of 2016,biggest celebrity scandals 2016,1
7,Caitlyn Jenner Addresses Rumored Romance With Sophia Hutchins,caitlyn jenner addresses rumored romance sophia hutchins,1
8,Taylor Swift Reportedly Reacts To Tom Hiddleston’s Golden Globes Win,taylor swift reportedly reacts tom hiddleston’s golden globes win,1
9,"For The Love Of God, Why Can't Anyone Write Kate McKinnon A Good Movie Role?",love god cant anyone write kate mckinnon good movie role,1


Now we have fake and real titles in order. In order to do train test split part as we want, we need to mix these data. And we need to delete repeated news, because we do not want our model to memorise patterns.

In [ ]:
df['clean_title'].replace(r'^\s*$', np.nan, regex=True, inplace=True) # replacing empty rows with NaN , and using regex (regular expression)

/tmp/ipython-input-4253311957.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['clean_title'].replace(r'^\s*$', np.nan, regex=True, inplace=True) # replacing empty rows with NaN , and using regex (regular expression)


In [ ]:
null_count = df['clean_title'].isnull().sum()
print(f"Number of empty rows after cleaning: {null_count}")

Number of empty rows after cleaning: 4


Here, we have 4 empty rows after second cleaning.

In [ ]:
if null_count > 0:
    df.dropna(subset=['clean_title'], inplace=True)
    print("Null Rows (empty titles) are dropped with dropna")

Null Rows (empty titles) are dropped with dropna


In [ ]:
# to eliminate empty rows

empty_rows = df[df['clean_title'].str.strip() == ""].index
print(f"Number of empty Rows: {len(empty_rows)}")

Number of empty Rows: 0


In [ ]:
# to mix our dataframe
df = df.sample(frac=1, random_state=42).reset_index(drop=True)
print("Dataframe is mixed up.")

Dataframe is mixed up.


Now we have fake and real titles in order. In order to do train test split part as we want, we need to mix these data. And we need to delete repeated news, because we do not want our model to memorise patterns.

In [ ]:
duplicate_count = df.duplicated(subset=['clean_title']).sum()
print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 1497


In [ ]:
if duplicate_count > 0:
    df.drop_duplicates(subset=['clean_title'], keep='first', inplace=True)
    print("Repeated news are dropped.")

 -> Repeated news dropped.


In [ ]:
df = df.sample(frac=1, random_state=42).reset_index(drop=True)
print("Dataframe is shuffled and index reset.")

Dataframe shuffled and index reset.


In [ ]:
df.to_csv("cleaned_news.csv", index=False)
print("Final CSV saved: cleaned_news.csv")

Final CSV saved: cleaned_news.csv


In [ ]:
df_cleaned = pd.read_csv("cleaned_news.csv")
print("\nFinal Data Info:")
df_cleaned.info()


Final Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23192 entries, 0 to 23191
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           23192 non-null  object
 1   news_url     22862 non-null  object
 2   title        23192 non-null  object
 3   tweet_ids    21692 non-null  object
 4   label        23192 non-null  int64 
 5   clean_title  23192 non-null  object
dtypes: int64(1), object(5)
memory usage: 1.1+ MB
